R/resource.R

```diff
@@ -0,0 +1,160 @@
+##'   nodes, then specify this here as a vector of strings for the node names.
+##'
+##' @param priority If the tasks you are launching are low priority, you can
+##'   allow other queuing tasks to jump over them, by setting the priority to
+##'   to `low`; otherwise, the default is `normal`. These are the only
+##'   acceptable values.
+##'
+##' @param queue Specify a particular queue to submit your tasks to.
+##'   This is in development as we decide over time what
+##'   queues we best need for DIDE's common workflows. See the Details
+##'   for more information
+##'
+##' @return A hipercow_resource list, which can be provided when you submit
+##'   tasks.
+##'
+##' @inheritParams hipercow_configure
+##'
+##' @export
+
+task_resources <- function(cores = 1L,
+                           exclusive = FALSE,
+                           runtime = NULL,
+                           hold_until = NULL,
+                           memory_per_node = NULL,
+                           memory_per_process = NULL,
+                           requested_nodes = NULL,
+                           priority = NULL,
+                           queue = NULL,
+                           driver = NULL,
+                           root = NULL) {
+  root <- hipercow_root(root)
+  dat <- hipercow_driver_prepare(driver, root, rlang::current_env())
+
+  validate_cores(cores)
+  if (rlang::is_integerish(cores)) {
+    cores <- as.integer(cores)
+  }
+
+  assert_scalar_logical(exclusive)
+  is.null(runtime) || validate_runtime(runtime)
+  is.null(hold_until) || validate_hold_until(hold_until)
+  is.null(memory_per_node) || validate_memory(memory_per_node)
+  is.null(memory_per_process) || validate_memory(memory_per_process)
+  is.null(requested_nodes) || validate_nodes(requested_nodes)
+  is.null(priority) || validate_priority(priority)
+  is.null(queue) || validate_queue(queue)
+
+  res <- as.list(environment())
+  class(res) <- "hipercow_resource"
+
+  res <- dat$driver$task_resources(res)
+  invisible(res)
+}
+
+validate_cores <- function(cores) {
+  assert_scalar(cores)
+  if (cores != Inf) {
+    if (!rlang::is_integerish(cores)) {
+      stop("Number of cores must be an integer, or Inf.")
+    }
+    if (cores <= 0) {
+      stop("Number of cores must be positive")
+    }
+  }
+}
+
+validate_runtime <- function(runtime) {
+  assert_scalar(runtime)
+  if (is.integer(runtime)) return()
+  stop("Runtime format failed")
+  #TODO
+  TRUE
+}
+
+validate_hold_until <- function(hold_until) {
+  if (is.character(hold_until)) {
+    if (hold_until %in% c("tonight", "midnight")) return()
+  }
+  #TODO
+  TRUE
+}
+
+validate_memory <- function(mem) {
+  assert_scalar(mem)
+  if (rlang::is_integerish(mem)) {
+    mem <- paste0(mem, "Gb")
+  }
+  gb <- grepl("^(\\d+)Gb$", mem)
+  tb <- grepl("^(\\d+)Tb$", mem)
+  if ((!gb) && (!tb)) {
+    cli::cli_abort(c(
+      "Couldn't intepret memory format from `{mem}`.",
+      i = "Examples: 10Gb or 1Tb"))
+  }
+  TRUE
```

Comment:
returning an integer here, as suggested in the longer comment above, will prevent this logic being duplicated everywhere